File: tests/parsing/java/qualified_super.java
// Test Java's qualified super syntax used in diamond inheritance

// 1. Qualified super method invocation (Type.super.method())
interface Left {
    default String getData() {
        return "left";
    }
}

interface Right {
    default String getData() {
        return "right";
    }
}

class Diamond implements Left, Right {
    @Override
    public String getData() {
        // Qualified super disambiguates which interface's default method to call
        return Left.super.getData();
    }

    public String getOther() {
        return Right.super.getData();
    }
}

// 2. Qualified super field access (Type.super.field)
class Parent {
    protected int value = 10;
}

class Child extends Parent {
    protected int value = 20;

    class Inner {
        int getValue() {
            // Access outer class's parent field
            return Child.super.value;
        }
    }
}

// 3. Qualified super constructor invocation (outer.super())
class Outer {
    class Inner {
    }
}

class ExtendedInner extends Outer.Inner {
    ExtendedInner(Outer outer) {
        // Explicit constructor invocation with qualified super
        outer.super();
    }
}

File: tests/parsing/scala/qualified_super.scala
// Test Scala's qualified super syntax

// 1. Qualified super with mixin type: super[Type].member
trait Left {
    def getData(): String = "left"
}

trait Right {
    def getData(): String = "right"
}

class Diamond extends Left with Right {
    override def getData(): String = {
        // Qualified super disambiguates which trait's implementation to call
        super[Left].getData()
    }

    def getOther(): String = {
        super[Right].getData()
    }
}

// 2. Prefixed super without mixin qualifier: Outer.super.member
class Parent {
    def getValue(): Int = 10
}

class Child extends Parent {
    override def getValue(): Int = 20

    class Inner {
        def getParentValue(): Int = {
            // Access outer class's parent implementation
            Child.super.getValue()
        }
    }
}

// 3. combine both of the above: Outer.super[Type].member
class Outer extends Left with Right {
    // Disambiguate the conflicting `getData` implementations from Left/Right.
    override def getData(): String = super[Left].getData()

    class Inner {
        def getOuterLeft(): String = {
            // Access outer class's Left trait implementation
            Outer.super[Left].getData()
        }

        def getOuterRight(): String = {
            Outer.super[Right].getData()
        }
    }
}